File: app/domain/comparison.py
from decimal import Decimal

from app import constants
from app.domain.models import PriceComparison


def compare_prices(
    sustainable_price: Decimal, expected_price: Decimal
) -> PriceComparison:
    """Compare the expected price against the sustainable price."""

    difference_percentage = (
        (expected_price - sustainable_price) / sustainable_price * Decimal("100")
    )

    if difference_percentage <= Decimal("-20"):
        return PriceComparison(
            difference_percentage=difference_percentage,
            category=constants.CATEGORY_VERY_BELOW,
            justification=constants.JUSTIFICATION_VERY_BELOW,
            color=constants.COLOR_RED,
        )

    if difference_percentage < Decimal("-5"):
        return PriceComparison(
            difference_percentage=difference_percentage,
            category=constants.CATEGORY_SLIGHTLY_BELOW,
            justification=constants.JUSTIFICATION_SLIGHTLY_BELOW,
            color=constants.COLOR_YELLOW,
        )

    if difference_percentage <= Decimal("5"):
        return PriceComparison(
            difference_percentage=difference_percentage,
            category=constants.CATEGORY_VERY_CLOSE,
            justification=constants.JUSTIFICATION_VERY_CLOSE,
            color=constants.COLOR_GREEN,
        )

    return PriceComparison(
        difference_percentage=difference_percentage,
        category=constants.CATEGORY_ABOVE,
        justification=constants.JUSTIFICATION_ABOVE,
        color=constants.COLOR_BLUE,
    )

File: src/domain/comparison.ts
import {
  CATEGORY_ABOVE,
  CATEGORY_SLIGHTLY_BELOW,
  CATEGORY_VERY_BELOW,
  CATEGORY_VERY_CLOSE,
  COLOR_BLUE,
  COLOR_GREEN,
  COLOR_RED,
  COLOR_YELLOW,
  JUSTIFICATION_ABOVE,
  JUSTIFICATION_SLIGHTLY_BELOW,
  JUSTIFICATION_VERY_BELOW,
  JUSTIFICATION_VERY_CLOSE,
} from "@/constants/constants";
import type { PriceComparison } from "@/types/pricing";

export function comparePrices(
  expectedPrice: number,
  sustainablePrice: number,
): PriceComparison {
  const differencePercentage =
    ((expectedPrice - sustainablePrice) / sustainablePrice) * 100;

  if (differencePercentage <= -20) {
    return {
      differencePercentage,
      category: CATEGORY_VERY_BELOW,
      justification: JUSTIFICATION_VERY_BELOW,
      color: COLOR_RED,
    };
  }

  if (differencePercentage < -5) {
    return {
      differencePercentage,
      category: CATEGORY_SLIGHTLY_BELOW,
      justification: JUSTIFICATION_SLIGHTLY_BELOW,
      color: COLOR_YELLOW,
    };
  }

  if (differencePercentage <= 5) {
    return {
      differencePercentage,
      category: CATEGORY_VERY_CLOSE,
      justification: JUSTIFICATION_VERY_CLOSE,
      color: COLOR_GREEN,
    };
  }

  return {
    differencePercentage,
    category: CATEGORY_ABOVE,
    justification: JUSTIFICATION_ABOVE,
    color: COLOR_BLUE,
  };
}
